[macOS] Fix rendering of <select> in vertical writing mode

pxlcoder opened the pull request:
#### 2efb1a7810e0576317097634354b4c2e942d9708
<pre>
[macOS] Fix rendering of &lt;select&gt; in vertical writing mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=264457">https://bugs.webkit.org/show_bug.cgi?id=264457</a>
<a href="https://rdar.apple.com/118147979">rdar://118147979</a>

Reviewed by Tim Nguyen.

`NSPopUpButton` is not height resizable, resulting in a squished appearance
for `&lt;select&gt;` in vertical writing mode. To fix, fall back to the
`MenuListButton` path and use custom rendering when using a vertical writing
mode. Update the painting code for menu list buttons to support vertical
writing mode.

* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
(WebCore::MenuListButtonMac::draw):

Remove incorrect logic to draw a separator next to the arrows. Currently, and
dating back to at least Safari 13, the separators do not show up (even in
horizontal writing mode). Fixing the logic to get separators working does not
result in an improvement in appearance. As a result, rather than re-writing the
logic, simply remove it.

* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::isControlStyled const):

Apply our fallback for zoomed `&lt;select&gt;` to `&lt;select&gt;`s that have a
vertical writing mode. In both cases, the fallback is necessary as the
native control only renders at fixed sizes.

Canonical link: <a href="https://commits.webkit.org/270458@main">https://commits.webkit.org/270458@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f4ae11d5db132a37e2cfec0ba911b41a6e0c784d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25509 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4102 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26785 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27613 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23379 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25789 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5842 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1537 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23533 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25755 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3034 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21989 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28194 "Built successfully") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22939 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29040 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23264 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23314 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26870 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2680 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/934 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4061 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6128 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3135 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3018 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/RenderThemeMac.mm:
```diff
@@ -759,11 +759,12 @@ static Color activeButtonTextColor()
-    // is in effect we treat it like the control is styled.
-    if (appearance == StyleAppearance::Menulist && style.effectiveZoom() != 1.0f)
+    // is in effect we treat it like the control is styled. Additionally, treat the control like it is styled when
+    // using a vertical writing mode, since menu lists are not height resizable.
```
since the AppKit appearance is not height resizable.?

pxlcoder replied:
Updated.

nt1m commented on Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
```diff
@@ -159,50 +162,38 @@ static void drawMenuListBackground(GraphicsContext& context, const FloatRect& re
+            return point.transposedPoint();
+        };
 
-    // FIXME: Should the separator thickness and space be scaled up by fontScale?
-    int separatorSpace = 2; // Deliberately ignores zoom since it looks nicer if it stays thin.
-    int leftEdgeOfSeparator;
-    if (isRightToLeft)
-        leftEdgeOfSeparator = static_cast<int>(roundf(leftEdge + arrowWidth + arrowPaddingBefore * style.zoomFactor));
-    else
-        leftEdgeOfSeparator = static_cast<int>(roundf(leftEdge - arrowPaddingBefore * style.zoomFactor));
-
-    // Draw the separator to the left of the arrows
-    context.setStrokeThickness(1); // Deliberately ignores zoom since it looks nicer if it stays thin.
-    context.setStrokeStyle(StrokeStyle::SolidStroke);
-    context.setStrokeColor(leftSeparatorColor);
-    context.drawLine(IntPoint(leftEdgeOfSeparator, bounds.y()), IntPoint(leftEdgeOfSeparator, bounds.maxY()));
```
thank you for getting rid of the invisible separator 😄

nt1m commented on Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
```diff
@@ -159,50 +162,38 @@ static void drawMenuListBackground(GraphicsContext& context, const FloatRect& re
-    else
-        leftEdgeOfSeparator = static_cast<int>(roundf(leftEdge - arrowPaddingBefore * style.zoomFactor));
-
-    // Draw the separator to the left of the arrows
-    context.setStrokeThickness(1); // Deliberately ignores zoom since it looks nicer if it stays thin.
-    context.setStrokeStyle(StrokeStyle::SolidStroke);
-    context.setStrokeColor(leftSeparatorColor);
-    context.drawLine(IntPoint(leftEdgeOfSeparator, bounds.y()), IntPoint(leftEdgeOfSeparator, bounds.maxY()));
+        topArrow = topArrow.map(transposePoint);
+        bottomArrow = bottomArrow.map(transposePoint);
+    }
```
Is this rotating the arrow? If it is, is it correct to rotate it? Does the popup appear where the arrow points?

nt1m replied:
Would be good to add a comment regardless.

pxlcoder replied:
Yes, this rotates the arrows – I've added a comment. 

The positioning of the menu isn't perfect, but it does appear to the side of the control (which is what we want) rather than above or below, hence the rotation.

nt1m commented on Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
```diff
@@ -159,50 +162,39 @@ static void drawMenuListBackground(GraphicsContext& context, const FloatRect& re
 
-    constexpr auto leftSeparatorColor = Color::black.colorWithAlphaByte(40);
-    constexpr auto rightSeparatorColor = Color::white.colorWithAlphaByte(40);
+    // Rotate the arrows for vertical writing mode.
```
nitpick, but comments usually mention _why_ we do things.

```suggestion
    // Rotate the arrows for vertical writing mode since the popup appears on the side of the menulist instead of under it.
```

pxlcoder committed:
[macOS] Fix rendering of <select> in vertical writing mode
https://bugs.webkit.org/show_bug.cgi?id=264457
rdar://118147979

Reviewed by Tim Nguyen.

`NSPopUpButton` is not height resizable, resulting in a squished appearance
for `<select>` in vertical writing mode. To fix, fall back to the
`MenuListButton` path and use custom rendering when using a vertical writing
mode. Update the painting code for menu list buttons to support vertical
writing mode.

* Source/WebCore/platform/graphics/mac/controls/MenuListButtonMac.mm:
(WebCore::MenuListButtonMac::draw):

Remove incorrect logic to draw a separator next to the arrows. Currently, and
dating back to at least Safari 13, the separators do not show up (even in
horizontal writing mode). Fixing the logic to get separators working does not
result in an improvement in appearance. As a result, rather than re-writing the
logic, simply remove it.

* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::isControlStyled const):

Apply our fallback for zoomed `<select>` to `<select>`s that have a
vertical writing mode. In both cases, the fallback is necessary as the
native control only renders at fixed sizes.

Canonical link: https://commits.webkit.org/270458@main
